android/libraries/rib-router-navigator/src/test/kotlin/com/uber/rib/core/RouterAndStateTest.kt

```diff
@@ -0,0 +1,193 @@
+  fun willAttachToHost_shouldCallProvidedAttachCallback() {
+
+    val routerAndState = RouterAndState(state, attachTransition, detachTransition)
+
+    routerAndState.willAttachToHost(null, false)
+
+    verify(attachTransition)
+      .willAttachToHost(same(routerAndState.router), eq(null), same(state), eq(false))
+  }
+
+  @Test
+  fun willDetachFromHost_whenDetachTransactionIsNotCallback_shouldCallProvidedDetachCallback() {
+    val routerAndState = RouterAndState(state, attachTransition, detachTransition)
+
+    routerAndState.willDetachFromHost(null, false)
+
+    verify(detachTransition)
+      .willDetachFromHost(same(routerAndState.router), same(state), eq(null), eq(false))
+  }
+
+  @Test
+  fun willDetachFromHost_whenDetachTransactionIsCallback_shouldCallProvidedDetachCallback() {
+    val routerAndState = RouterAndState(state, attachTransition, detachCallback)
+
+    routerAndState.willDetachFromHost(null, false)
+
+    verify(detachCallback)
+      .willDetachFromHost(same(routerAndState.router), same(state), eq(null), eq(false))
+  }
+
+  @Test
+  fun willDetachFromHost_whenDetachTransactionIsNotCallback_shouldCallProvidedDetachTransaction() {
+    val routerAndState = RouterAndState(state, attachTransition, detachTransition)
+
+    val router1 = routerAndState.router
+    routerAndState.willDetachFromHost(null, false)
+
+    verify(detachTransition).willDetachFromHost(same(router1), same(state), eq(null), eq(false))
+  }
+
+  @Test
+  fun onPostDetachFromHost_shouldCallProvidedDetachCallback() {
+    val routerAndState = RouterAndState(state, attachTransition, detachCallback)
+
+    val router1 = routerAndState.router
+    routerAndState.onPostDetachFromHost(null, false)
+
+    verify(detachCallback).onPostDetachFromHost(same(router1), eq(null), eq(false))
+  }
+
+  @Test
+  fun router_whenCurrentlyDestroyingTheInstance_shouldCallBuilderTwice() {
+    val createdRouters = mutableSetOf<Router<*>>()
+    val destroyedRouters = mutableSetOf<Router<*>>()
+
+    whenever(attachTransition.buildRouter()).doAnswer {
+      mock<Router<*>>().apply(createdRouters::add)
+    }
+    whenever(detachCallback.onPostDetachFromHost(any(), anyOrNull(), any())).then {
+      destroyedRouters.add(it.arguments[0] as Router<*>)
+    }
+
+    val routerAndState =
+      RouterAndState(
+        state,
+        attachTransition,
+        detachCallback,
+      )
+
+    val threadsCount = 10
+    val finishedThreadsCount = AtomicInteger(0)
+
+    // Create 10 threads with 1000 routers usage flows inside each
+    repeat(threadsCount) {
+      thread {
+        repeat(1000) {
+          routerAndState.willAttachToHost(null, false)
+          routerAndState.willDetachFromHost(null, false)
+          routerAndState.onPostDetachFromHost(null, false)
+        }
+        finishedThreadsCount.addAndGet(1)
+      }
+    }
+
+    while (finishedThreadsCount.get() < threadsCount) {
+      // Waiting until all threads finish their work
+      Thread.sleep(10)
+    }
```

Comment:
A countdown latch is a better solution than randomly sleeping 10ms in a tight loop:
```kotlin
    val threadsCount = 10
    val latch = CountDownLatch(threadsCount)
    // Create 10 threads with 1000 routers usage flows inside each
    repeat(threadsCount) {
      thread {
        repeat(1000) {
          routerAndState.willAttachToHost(null, false)
          routerAndState.willDetachFromHost(null, false)
          routerAndState.onPostDetachFromHost(null, false)
        }
        latch.countDown()
      }
    }
    latch.await()
```

Reply:
Nice, didn't know that `CountDownLatch ` exist.